GLMakie/src/new-primitives.jl

```diff
@@ -147,19 +150,31 @@ end
 
 # TODO: Consider separating this from plot (i.e. update this in render using scene graph)
 function register_light_attributes!(screen, scene, attr, uniforms)
-    if !(haskey(attr, :shading) && attr[:shading][])
-        return
+    # plot does not support shading
+    haskey(attr, :shading) || return
+
+    # On re-display these are already registered. To allow compiling shaders
+    # with different light settings we need to clear old computations
+    if haskey(attr, :ambient)
+        if haskey(attr, :gl_renderobject)
+            error("Try to register lights that have already been registered. Is the renderobject getting created twice?")
+        else
+            for key in [:ambient, :light_color, :light_direction, :N_lights, :light_types, :light_colors, :light_parameters]
+                if haskey(attr, key)
+                    delete!(attr, key, force = true)
+                end
+            end
+        end
     end
 
+    # Nothing to generate if we don't shade
     shading = Makie.get_shading_mode(scene)
-    shading == NoShading && return
-    if !haskey(attr, :ambient)
-        # TODO, why is ambient already here for MeshScatter?
-        add_input!(attr, :ambient, scene.compute[:ambient_color]::Computed)
-    else
+    if !attr[:shading][] || (shading == NoShading)
```

Comment:
should this be `attr[:shading] == NoShading` via convert_attribute?

Reply:
No, I changed `plot.shading` to cast to a Bool, which is `attr[:shading]` here. The variable `shading` comes from the scene to decide the algorithm based on the lights in the scene.